python/tvm/relay/build_module.py

```diff
@@ -217,30 +229,147 @@ def build(func,
+
+    fallback_device : str or tvm.TVMContext, optional.
+        The fallback device. It is also used as the default device for
+        operators with no specified device.
+
+    Returns
+    -------
+    device_target : dict of int to str.
+        The updated device id to target dict.
+
+    fallback_device : int
+        The updated fallback device id.
+    """
+    if not isinstance(target, dict):
+        raise ValueError("target must be dict of device name to target for " +
+                         "heterogeneous execution, but received %s."
+                         % type(target))
+
+    if fallback_device is None:
+        # cpu is used as the default fallback device when heterogeneous
+        # execution is needed, but no fallback device is provided.
+        fallback_device = _nd.cpu(0).device_type
+        target[fallback_device] = str(_target.create("llvm"))
+    elif isinstance(fallback_device, str):
+        fallback_device = _nd.context(fallback_device).device_type
+    elif isinstance(fallback_device, TVMContext):
+        fallback_device = fallback_device.device_type
+    else:
+        raise ValueError("fallback_device expects the type of str or" +
+                         "TVMContext, but received %s." % type(fallback_device))
+
+    device_target = {}
+    for dev, tgt in target.items():
+        device_target[_nd.context(dev).device_type] = str(tgt)
+
+    if fallback_device not in device_target:
+        raise ValueError("%s is used as the default device, but the target" +
+                         "is not provided."
+                         % _nd.context(fallback_device).device_name)
+    return device_target, fallback_device
+
+
+def _run_device_annotation_passes(func, target, fallback_device):
+    """Execute the device annotation passes to update the input program and
+    target information.
+
+    Parameters
+    ----------
+    func: tvm.relay.Function
+        The function where annotation passes will be execute at.
+
+    target : Dict[int, tvm.target.Target]
+        A dict contains device_id to target pairs.
+
+    fallback_device : int
+        The fallback device type.
+
+    Returns
+    -------
+    target : Dict[int, tvm.target.Target]
+        The updated device id to target dict.
+
+    func : tvm.relay.Function
+        The updated func.
+    """
+    func = ir_pass.infer_type(func)
+    func = ir_pass.rewrite_annotated_ops(func, fallback_device)
+    device_map = ir_pass.collect_device_info(func)
+    # The expression to device id map will be empty if all or none of
+    # the expressions in the `func` are annotated because this map is
+    # obtained by propagating the device information in the device copy
+    # operator. None of the above cases needs device copy operator.
+    if not device_map:
+        annotation_map = ir_pass.collect_device_annotation_ops(func)
+        # No annotation.
+        if not annotation_map:
+            target = {0: target[fallback_device]}
```

Comment:
just to make sure, so `0` is a special `dev_id` which is not assigned to a device in current tvm code base, right?

Reply:
@yzhliu Yes, and it is also used in `graph_plan_memory`.

Reply:
also it is `DLDeviceType`, so should be "device_type", right?